File: n2/5_2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstdlib>
#include <algorithm>

using namespace std;

string ticket() {
    int num = rand() % 1000;
    string t = "T";
    if (num < 10) t += "00";
    else if (num < 100) t += "0";
    t += to_string(num);
    return t;
}

int main() {
    int n;
    cout << "Доступные команды: ENQUEUE, DISTRIBUTE\n";
    cout << "Введите количество окон: ";
    cin >> n;

    vector<pair<string, int>> tasks;
    string command;

    while (true) {
        cout << "Команда: ";
        cin >> command;

        if (command == "ENQUEUE") {
            int time;
            cin >> time;
            string tick = ticket();
            tasks.push_back(make_pair(tick, time));
            cout << tick << endl;
        }
        else if (command == "DISTRIBUTE") {
            sort(tasks.begin(), tasks.end(), [](const pair<string, int>& a, const pair<string, int>& b) {
                return a.second > b.second;
            });

            vector<vector<string>> window_queue(n);
            vector<int> window_time(n, 0);

            for (size_t i = 0; i < tasks.size(); ++i) {
                const string& tick = tasks[i].first;
                int time = tasks[i].second;
                int min_index = 0;
                for (int j = 1; j < n; ++j) {
                    if (window_time[j] < window_time[min_index])
                        min_index = j;
                }

                window_queue[min_index].push_back(tick);
                window_time[min_index] += time;
            }

            for (int i = 0; i < n; ++i) {
                cout << "Окно " << i + 1 << " (" << window_time[i] << " минут): ";
                for (size_t j = 0; j < window_queue[i].size(); ++j) {
                    cout << window_queue[i][j];
                    if (j < window_queue[i].size() - 1) cout << ", ";
                }
                cout << "\n";
            }
            break;
        }
        else {
            cout << "Неизвестная команда" << endl;
        }
    }

    return 0;
}
//

File: n2/5_2.rs
use rand::Rng;
use std::cmp::Ordering;
use std::collections::VecDeque;
use std::io::{self, Write};

fn ticket() -> String {
    let num = rand::thread_rng().gen_range(0..1000);
    format!("T{:03}", num)
}

fn main() {
    println!("Доступные команды: ENQUEUE, DISTRIBUTE");
    print!("Введите количество окон: ");
    io::stdout().flush().unwrap();

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let n: usize = input.trim().parse().unwrap();

    let mut tasks: Vec<(String, i32)> = Vec::new();

    loop {
        print!("Команда: ");
        io::stdout().flush().unwrap();
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let command = input.trim();

        if command == "ENQUEUE" {
            print!("Введите время: ");
            io::stdout().flush().unwrap();
            input.clear();
            io::stdin().read_line(&mut input).unwrap();
            let time: i32 = input.trim().parse().unwrap();
            let tick = ticket();
            tasks.push((tick.clone(), time));
            println!("{}", tick);
        } else if command == "DISTRIBUTE" {
            tasks.sort_by(|a, b| b.1.cmp(&a.1));

            let mut window_queues: Vec<Vec<String>> = vec![vec![]; n];
            let mut window_times: Vec<i32> = vec![0; n];

            for (tick, time) in &tasks {
                let min_index = window_times
                    .iter()
                    .enumerate()
                    .min_by_key(|&(_, &v)| v)
                    .map(|(i, _)| i)
                    .unwrap();

                window_queues[min_index].push(tick.clone());
                window_times[min_index] += time;
            }

            for i in 0..n {
                print!("Окно {} ({} минут): ", i + 1, window_times[i]);
                for (j, t) in window_queues[i].iter().enumerate() {
                    print!("{}", t);
                    if j < window_queues[i].len() - 1 {
                        print!(", ");
                    }
                }
                println!();
            }
            break;
        } else {
            println!("Неизвестная команда");
        }
    }
}